### PR TITLE
Added BORG_PASSCOMMAND environment variable

### PR DESCRIPTION
**Pull request by TuXicc**
The BORG_PASSCOMMAND variable can be used to specify an external program that is asked to supply the passphrase.

### CODE REVIEW

File: src/borg/crypto/key.py [hunk 1]
```diff
@@ -406,18 +410,33 @@ def init_ciphers(self, manifest_nonce=0):
...
+        passphrase = cls._get_env('BORG_PASSPHRASE', default)
+        if passphrase is not None:
+            return passphrase
+        passphrase = cls.env_passcommand()
+        if passphrase is not None:
+            return passphrase
+
+    @classmethod
+    def env_passcommand(cls, default=None):
+        passcommand = cls._get_env('BORG_PASSCOMMAND', default)
```
Comment (ThomasWaldmann) [inline src/borg/crypto/key.py]: that's a bit weird. the returned object is of class `Passphrase` (if not None), but does contain the command, not a passphrase. so i guess you'ld rather call `os.environ.get('BORG_PASSCOMMAND', default)` here.

Comment (ThomasWaldmann) [inline src/borg/crypto/key.py]: ... and when doing it like that, you maybe do not need to rename _env_passphrase method.

File: src/borg/crypto/key.py [hunk 2]
```diff
@@ -406,18 +410,33 @@ def init_ciphers(self, manifest_nonce=0):
...
+        if passcommand is not None:
+            try:
+                passphrase = subprocess.check_output(passcommand, shell=True, universal_newlines=True)
+            except subprocess.CalledProcessError as e:
+                raise PasscommandFailure(e)
+            return Passphrase(passphrase.rstrip('\n'))
```
Comment (ThomasWaldmann) [inline src/borg/crypto/key.py]:
```
return cls(...)
```

File: docs/usage_general.rst.inc [hunk 3]
```diff
@@ -131,12 +131,19 @@ General:
         can either leave it away or abbreviate as `::`, if a positional parameter is required.
     BORG_PASSPHRASE
         When set, use the value to answer the passphrase question for encrypted repositories.
-        It is used when a passphrase is needed to access a encrypted repo as well as when a new
+        It is used when a passphrase is needed to access an encrypted repo as well as when a new
         passphrase should be initially set when initializing an encrypted repo.
         See also BORG_NEW_PASSPHRASE.
+    BORG_PASSCOMMAND
+        When set, use the command to answer the passphrase question for encrypted repositories.
```
Comment (ThomasWaldmann) [inline docs/usage_general.rst.inc]: ... via the standard output of the command (trailing newlines are stripped).

File: src/borg/crypto/key.py (hunk 1 above)
Comment (enkore) [inline src/borg/crypto/key.py]: os.environ.get(..., None); the passed `default` is meant as the default for the passphrase

File: src/borg/crypto/key.py [hunk 4]
```diff
@@ -406,18 +410,33 @@ def init_ciphers(self, manifest_nonce=0):
...
+        passphrase = cls.env_passcommand()
+        if passphrase is not None:
+            return passphrase
+
+    @classmethod
+    def env_passcommand(cls, default=None):
+        passcommand = cls._get_env('BORG_PASSCOMMAND', default)
+        if passcommand is not None:
+            try:
+                passphrase = subprocess.check_output(passcommand, shell=True, universal_newlines=True)
```
Comment (enkore) [inline src/borg/crypto/key.py]: I'd prefer to avoid shell=True

File: src/borg/crypto/key.py (hunk 1 above)
Comment (ThomasWaldmann) [inline src/borg/crypto/key.py]: right, what @enkore says.

**Comment by enkore** [comment]
Thanks!

**Comment by enkore** [comment]
Stapling #392

**Pull request merged**